gdal/frmts/gtiff/tif_jxl.c

```diff
@@ -0,0 +1,826 @@
+            JxlDecoderReleaseInput(sp->decoder);
+            return 0;
+        }
+
+        status = JxlDecoderProcessInput(sp->decoder);
+        if( status != JXL_DEC_FULL_IMAGE )
+        {
+            TIFFErrorExt(tif->tif_clientdata, module,
+                         "JxlDecoderProcessInput() (third call) failed with %d", status);
+            JxlDecoderReleaseInput(sp->decoder);
+            return 0;
+        }
+
+        /*const size_t nRemaining = */ JxlDecoderReleaseInput(sp->decoder);
+        /*if( nRemaining != 0 )
+        {
+            TIFFErrorExt(tif->tif_clientdata, module,
+                         "JxlDecoderReleaseInput(): %u input bytes remaining",
+                         (unsigned)nRemaining);
+        }*/
+
+        return 1;
+}
+
+/*
+* Decode a strip, tile or scanline.
+*/
+static int
+JXLDecode(TIFF* tif, uint8_t* op, tmsize_t occ, uint16_t s)
+{
+        static const char module[] = "JXLDecode";
+        JXLState* sp = DecoderState(tif);
+
+        (void) s;
+        assert(sp != NULL);
+        assert(sp->state == LSTATE_INIT_DECODE);
+
+        if( sp->uncompressed_buffer == 0 )
+        {
+            TIFFErrorExt(tif->tif_clientdata, module,
+                         "Uncompressed buffer not allocated");
+            return 0;
+        }
+
+        if( (uint64_t)sp->uncompressed_offset +
+                                        (uint64_t)occ > sp->uncompressed_size )
+        {
+            TIFFErrorExt(tif->tif_clientdata, module,
+                         "Too many bytes read");
+            return 0;
+        }
+
+        memcpy(op,
+               sp->uncompressed_buffer + sp->uncompressed_offset,
+               occ);
+        sp->uncompressed_offset += (unsigned)occ;
+
+        return 1;
+}
+
+static int
+JXLSetupEncode(TIFF* tif)
+{
+        JXLState* sp = EncoderState(tif);
+
+        assert(sp != NULL);
+        if (sp->state & LSTATE_INIT_DECODE) {
+            sp->state = 0;
+        }
+
+        sp->state |= LSTATE_INIT_ENCODE;
+
+        return 1;
+}
+
+/*
+* Reset encoding state at the start of a strip.
+*/
+static int
+JXLPreEncode(TIFF* tif, uint16_t s)
+{
+        static const char module[] = "JXLPreEncode";
+        JXLState *sp = EncoderState(tif);
+        TIFFDirectory *td = &tif->tif_dir;
+
+        (void) s;
+        assert(sp != NULL);
+        if( sp->state != LSTATE_INIT_ENCODE )
+            tif->tif_setupencode(tif);
+
+        if( td->td_planarconfig == PLANARCONFIG_CONTIG &&
+            td->td_samplesperpixel > 4 )
+        {
+            TIFFErrorExt(tif->tif_clientdata, module,
+                "JXL: Only up to 4 samples are supported for Planarconfig=Contig");
+            return 0;
+        }
+
+        if( !sp->lossless && td->td_planarconfig == PLANARCONFIG_CONTIG )
+        {
+            switch(td->td_samplesperpixel) {
+                case 2:
+                    if(td->td_sampleinfo[1] != EXTRASAMPLE_UNASSALPHA) {
```

Comment:
td_sampleinfo[] is not necessarily td_samplesperpixel large:
```suggestion
                    if( !(td->td_extrasamples > 0 && td->td_sampleinfo[td->td_extrasamples-1] == EXTRASAMPLE_UNASSALPHA) ) {
```